BarrysPeanutsTemporalJava/src/main/java/barryspeanuts/ShoppingCartWorkflowImpl.java

```diff
@@ -0,0 +1,95 @@
+package barryspeanuts;
+
+import barryspeanuts.model.Purchase;
+import barryspeanuts.model.PurchaseItem;
+import barryspeanuts.task.CheckOutTaskImpl;
+import barryspeanuts.task.PayTaskImpl;
+import barryspeanuts.task.ShipTaskImpl;
+import barryspeanuts.task.WorkflowTask;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.WorkflowQueue;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import org.slf4j.Logger;
+
+public class ShoppingCartWorkflowImpl implements ShoppingCartWorkflow {
+  private static final Logger logger = Workflow.getLogger(ShoppingCartWorkflowImpl.class);
+  ShoppingCartActivities activities = new ShoppingCartActivitiesImpl();
+  List<PurchaseItem> purchaseItems;
+  private WorkflowQueue<WorkflowTask> queue = Workflow.newWorkflowQueue(1024);
+
+  @Override
+  public void startWorkflow() {
+    logger.info("Starting Workflow for Barry's Peanuts");
+    while (true) {
+      if (Workflow.getInfo().getHistoryLength() > 2000) {
+        // do continue as new here
+        logger.info("Workflow history greater than 2000");
+        // System.out.println("Workflow history greater than 2000");
+      }
+
+      WorkflowTask task = this.queue.cancellablePoll(Duration.ofDays(30));
+      if (task == null) {
+        logger.info("Breaking task");
+        break;
+      }
+      task.process(this);
+    }
+  }
+
+  @Override
+  public ShoppingCartActivities queryActivities() {
+    return this.activities;
+  }
+
+  @Override
+  public List<PurchaseItem> queryPurchaseItems() {
+    return this.purchaseItems;
+  }
+
+  @Override
+  public void addItem(PurchaseItem purchaseItem) {
+    if (this.purchaseItems == null) {
+      this.purchaseItems = new ArrayList<>();
+    }
+    this.purchaseItems.add(purchaseItem);
+  }
+
+  @Override
+  public void removeItem(PurchaseItem purchaseItem) {
+    if (this.purchaseItems != null) {
+      this.purchaseItems.remove(purchaseItem);
+    }
+  }
+
+  public void clearItems() {
+    this.purchaseItems = new ArrayList<>();
+  }
+
+  @Override
+  public void checkOut(String message) {
+    this.queue.put(new CheckOutTaskImpl(this.getPurchase()));
```

Comment:
I think you're overthinking/over-abstracting the concept of tasks and it's requiring a whole new package and several Java files. I can understand why you don't want to execute the activity right here, but can implement most of this inside the workflow as just functions. Just make the queue accept callables or something.

The goal of the whole project (and most projects really) should be simplicity. There's no need to fan out to lots of classes when simple code will do.

Reply:
OK, @cretz  Let me rework the code and see what I come up with. Can you elaborate on what you mean by "Just make the queue accept callables or something"

Reply:
Runnable makes more sense callable, so something like:

```java
private WorkflowQueue<Runnable> queue = Workflow.newWorkflowQueue(1024);
```

Then:

```java
  @Override
  public void checkOut(String message) {
    queue.put(() -> processCheckout(message));
  }

  private void processCheckout(String message) {
    // ... 
  }
```

Reply:
Thanks! I'll do as you advise.

Reply:
@cretz I did the refactoring. I am sure there are still opportunities for improvement and correction. Should I nuke this PR and create a new one so you have a clean slate to comment against?